src/types.rs

```diff
@@ -11,3 +14,32 @@ impl Mode {
         self.0
     }
 }
+
+#[repr(transparent)]
+pub struct CStr(str);
+
+impl CStr {
+    pub fn new<'a>(data: &'a str) -> &'a CStr {
+        if data.bytes().position(|b| b == b'\x00') != Some(data.len() - 1) {
+            panic!("CStr must contain a single NUL byte at the end");
```

Comment:
Hm, is it worth having two versions of this, one internal for `cstr!`(I think it needs to be `#[doc(hidden)] pub`) and one actually public that returns EINVAL instead of panicking?

Reply:
As soon as we have another use case for this :-)

Also, cstr doesn't have a way to statically verify these things, so it can't  bypass them at runtime either.

Reply:
TBH I'd be okay with this being an unsafe function that assumes there's a trailing NUL, to get rid of the runtime cost. `cstr!` enforces it so it can safely call it.